### PR TITLE
Feature Branch: merge master to `planner_refactory` branch

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #4739 
close #5350 

Problem Summary:

### What is changed and how it works?

We can use https://github.com/pingcap/tiflash/pull/5350 to check if the modification is correct

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Planner/Planner.cpp [hunk 1]
```diff
@@ -87,9 +87,15 @@ BlockInputStreams Planner::execute()
...
+    /// todo support fine grained shuffle
+    static auto disable_fine_frained_shuffle = [](const DAGQueryBlock & query_block) {
+        return !enableFineGrainedShuffle(query_block.source->fine_grained_shuffle_stream_count())
+            && (!query_block.exchange_sender || !enableFineGrainedShuffle(query_block.exchange_sender->fine_grained_shuffle_stream_count()));
+    };
     return query_block.source
         && (query_block.source->tp() == tipb::ExecType::TypeProjection
-            || query_block.source->tp() == tipb::ExecType::TypeExchangeReceiver);
+            || query_block.source->tp() == tipb::ExecType::TypeExchangeReceiver)
+        && disable_fine_frained_shuffle(query_block);
```
Comment (SeaRise) [inline dbms/src/Flash/Planner/Planner.cpp:98]: for fine grained shuffle
https://github.com/pingcap/tiflash/pull/5048

File: dbms/src/Flash/Planner/plans/PhysicalExchangeReceiver.cpp [hunk 2]
```diff
@@ -69,7 +68,7 @@ void PhysicalExchangeReceiver::transformImpl(DAGPipeline & pipeline, Context & c
     auto & exchange_receiver_io_input_streams = dag_context.getInBoundIOInputStreamsMap()[executor_id];
     for (size_t i = 0; i < max_streams; ++i)
     {
-        BlockInputStreamPtr stream = std::make_shared<ExchangeReceiverInputStream>(mpp_exchange_receiver, log->identifier(), executor_id);
+        BlockInputStreamPtr stream = std::make_shared<ExchangeReceiverInputStream>(mpp_exchange_receiver, log->identifier(), executor_id, /*stream_id=*/0);
```
Comment (SeaRise) [inline dbms/src/Flash/Planner/plans/PhysicalExchangeReceiver.cpp:71]: for fine grained shuffle
https://github.com/pingcap/tiflash/pull/5048

File: dbms/src/Flash/Planner/plans/PhysicalExchangeSender.cpp [hunk 3]
```diff
@@ -55,18 +55,21 @@ void PhysicalExchangeSender::transformImpl(DAGPipeline & pipeline, Context & con
...
             partition_col_ids,
             partition_col_collators,
             exchange_type,
             context.getSettingsRef().dag_records_per_chunk,
             context.getSettingsRef().batch_send_min_limit,
             stream_id++ == 0, /// only one stream needs to sending execution summaries for the last response
-            dag_context);
+            dag_context,
+            0,
+            0);
```
Comment (SeaRise) [inline dbms/src/Flash/Planner/plans/PhysicalExchangeSender.cpp:72]: for fine grained shuffle
https://github.com/pingcap/tiflash/pull/5048

File: dbms/src/Flash/tests/gtest_planner_interpreter.cpp [hunk 4]
```diff
@@ -673,5 +397,548 @@ CreatingSets
...
+             Union: <for partial order>
+              PartialSorting x 10: limit = 0
+               Expression: <final projection>
+                MockTableScan)";
+        ASSERT_BLOCKINPUTSTREAM_EQAUL(expected, request, 10);
+    }
+}
+CATCH
+
+/// todo support FineGrainedShuffle
```
Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_planner_interpreter.cpp:628]: for fine grained shuffle
https://github.com/pingcap/tiflash/pull/5048

File: dbms/src/Flash/Planner/plans/PhysicalAggregation.cpp [hunk 5]
```diff
@@ -106,34 +106,39 @@ void PhysicalAggregation::transformImpl(DAGPipeline & pipeline, Context & contex
...
-            pipeline.streams.resize(1);
-        if (stream_with_non_joined_data)
-            inputs.push_back(stream_with_non_joined_data);
+
+        if (!pipeline.streams_with_non_joined_data.empty())
+            inputs.push_back(pipeline.streams_with_non_joined_data.at(0));
+
+        pipeline.streams.resize(1);
+        pipeline.streams_with_non_joined_data.clear();
+
```
Comment (SeaRise) [inline dbms/src/Flash/Planner/plans/PhysicalAggregation.cpp:141]: from https://github.com/pingcap/tiflash/pull/5274

**Review by ywqzzy — APPROVED** [review]

**Review by guo-shaoge — APPROVED** [review]

**Pull request merged**